### PR TITLE
nicholascassar/at 240/add payments

### PR DESCRIPTION
**Pull request by nscassar**
Payments functionality added. Note that `Balance` on the payment tab shows the combined balance for dairy transactions, loans, and payments. The `Balance` on the loans tab shows the sum of all loans, and the `Balance` on the collect tab shows the sum of all dairy transactions.

### CODE REVIEW

File: src/utils/CoreAPI/index.ts [hunk 1]
```diff
@@ -9,7 +9,13 @@ import { NetworkError } from '../../errors/NetworkError';
...
 
 /** Paths on Core for specific data tables */
-export type CorePath = '/people/farmers' | '/transactions/products/milk' | '/transactions/products/export' | '/transactions/money/loan';
+export type CorePath =
+  '/people/farmers' | 
+  '/transactions/products/milk' | 
+  '/transactions/products/export' | 
+  '/transactions/money/loan' | 
+  '/transactions/money/payment'
+;
```
Comment (mfalthaw) [inline src/utils/CoreAPI/index.ts]: Move `;` to the previous line.

File: src/view/components/AddPaymentEntry/index.tsx [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  private allValid = () => (
+    this.state.validAmount
+  )
+
+  /** 
+   * Checks when the amount field has something in it 
+   * and verifies that it is a are real and positive number
+   */
+  private onChangeAmount = (newAmount: string) => {
+    const newAmountInt = parseInt(newAmount, radix);
```
Comment (mfalthaw) [inline src/view/components/AddPaymentEntry/index.tsx]: Now that I think of it, this should be a `float` not `int`. Please parse is as `float` and change all the `AddEntry` components to do the same.

You could use this as a reference
https://www.w3schools.com/Jsref/jsref_parsefloat.asp

File: src/view/components/AddPaymentEntry/style.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { StyleSheet } from 'react-native';
+
+/** Add Loan Entry stylesheet */
```
Comment (mfalthaw) [inline src/view/components/AddPaymentEntry/style.ts]: `payment`*

File: src/view/components/EditPaymentEntry/style.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import { StyleSheet } from 'react-native';
+
+/** Add Loan Entry stylesheet */
```
Comment (mfalthaw) [inline src/view/components/EditPaymentEntry/style.ts]: `Edit Payment Entry`*

File: src/view/components/PaymentEntryDetails/style.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { StyleSheet } from 'react-native';
+
+/** Add Loan Entry stylesheet */
```
Comment (mfalthaw) [inline src/view/components/PaymentEntryDetails/style.ts]: `Payment Entry Details`*

File: src/view/components/EditPaymentEntry/index.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  private allValid = () => (
+    this.state.validAmount
+  )
+
+  /** 
+   * Checks when the amount field has something in it 
+   * and verifies that it is a are real and positive number
+   */
+  private onChangeAmount = (newAmount: string) => {
+    const newAmountInt = parseInt(newAmount, radix);
```
Comment (mfalthaw) [inline src/view/components/EditPaymentEntry/index.tsx]: Don't forget about parsing to float here.

File: src/view/navigation/routes.ts [hunk 7]
```diff
@@ -139,7 +145,8 @@ export const routesInfo: RouteInfo[] = [
   {
     route: Route.MILK_ENTRY_DETAILS,
     name: 'Milk Entry Details',
-    component: MilkEntryDetailsPage,
+    component: MilkEntryDetailsPage
+    ,
```
Comment (mfalthaw) [inline src/view/navigation/routes.ts]: Move comma back up.

File: src/view/navigation/routes.ts [hunk 8]
```diff
@@ -160,6 +167,24 @@ export const routesInfo: RouteInfo[] = [
...
     type: 'back',
   },
+  {
+    route: Route.ADD_PAYMENT_ENTRY,
+    name: 'Add Payment Entry',
+    component: AddPaymentEntryPage,
+    type: 'back',
+  },
+  {
+    route: Route.LOAN_ENTRY_DETAILS,
```
Comment (mfalthaw) [inline src/view/navigation/routes.ts]: This is why `Edit Entry` won't work, this should route to `EDIT_PAYMENT_ENTRY` not `LOAN_ENTRY_DETAILS`

File: src/utils/CoreAPI/index.ts (hunk 1 above)
Comment (nscassar) [reply src/utils/CoreAPI/index.ts]: Updated

File: src/view/components/AddPaymentEntry/style.ts (hunk 3 above)
Comment (nscassar) [reply src/view/components/AddPaymentEntry/style.ts]: Updated

File: src/view/components/EditPaymentEntry/style.ts (hunk 4 above)
Comment (nscassar) [reply src/view/components/EditPaymentEntry/style.ts]: Updated

File: src/view/navigation/routes.ts (hunk 7 above)
Comment (nscassar) [reply src/view/navigation/routes.ts]: Updated

File: src/view/navigation/routes.ts (hunk 8 above)
Comment (nscassar) [reply src/view/navigation/routes.ts]: That was terrible, I can't believe I missed that. Updated.

File: src/store/modules/payment/selectors.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+);
+
+/** 
+ * Selector for the farmer account balance 
+ * Note: This returns payments only from a week before the current date
+ */
+export const getFarmerPaymentBalance = createSelector(
+  [getFarmersTransactions],
+  (paymentEntries: PaymentEntry[]) => paymentEntries.reduce((sum: number, entry: PaymentEntry) =>
+  (inLastWeek(entry.datetime)) ? sum + entry.amount : sum + 0, 0).toString(),
```
Comment (jeinglis) [inline src/store/modules/payment/selectors.ts]: same as above include all payments in the calculation

File: src/store/selectors.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import { getFarmerPaymentBalance } from './modules/payment/selectors';
+
+/**
+ * Selector for the farmer total balance for the week
+ * Includes payments, loans, and dairy transactions
+ */
+export const getFarmerTotalBalance = createSelector(
+  [getFarmerPaymentBalance, getFarmerLoanBalance, getFarmerDairyBalance],
+  (paymentBalance: string, loanBalance: string, dairyBalance: string) => 
+  ((Number(dairyBalance) - Number(paymentBalance) - Number(loanBalance)).toString()),
```
Comment (jeinglis) [inline src/store/selectors.ts]: Looks good only problem I can see is that the decimal on the balance throws the page formatting off. could you implement that formatter selector function to truncate the decimals and add the K suffix to large numbers.

Comment (nscassar) [reply src/store/selectors.ts]: Absolutely

Comment (nscassar) [reply src/store/selectors.ts]: Implemented. I also tweaked the formatter to allow for large negative numbers as well.

**Pull request merged**